Post about security in rotki (#29)
rotkehlchenio/rotkicom.github.io: _posts/2026-05-21-rotki-security.md

---
layout: post
title: "How we shield rotki against supply chain attacks"

> description: with the increasing amount of attacks is important to ensure integrity and secure the whole developement pipeline, this is how we do it at rotki.

Reviewer: ```suggestion
description: With the increasing number of attacks, it is important to ensure integrity and secure the whole development pipeline. This is how we do it at rotki.
```

tags:
- rotki
- developer

keywords:
- rotki
- security
- privacy
- local fist
---

rotki is a local app that helps you to keep track of your crypto activity both as transactions on different chains and on CEXes like Coinbase, Kraken, Binance, etc. This information is quite sensitive and there are [reasons to not leak it](https://github.com/jlopp/physical-bitcoin-attacks/blob/master/README.md). It is **critical** for us that what reaches your computer is exactly what we built. Anything less puts our users at risk.

For this reason, after the recent attacks [^1] [^2] [^3], we reviewed our dependency supply chain and the measures we've had in place since day one. rotki has many moving parts and its architecture spans different stacks and layers: package managers, CI, packaging pipelines, distribution. All of them need to be secure.

> There is no magic solution that makes the process 100% secure. ut you can make attacks harder, reduce the places where trust is implicit, and make releases easier to verify. This is what we currently do.

Reviewer: ```suggestion
There is no magic solution that makes the process 100% secure. But you can make attacks harder, reduce the places where trust is implicit, and make releases easier to verify. This is what we currently do.
```

With locked installs, CI installs what we reviewed before. If someone wants to update a package, that becomes an explicit code change.

Also for the `node` ecosystem we use `pnpm` instead of `npm` since it offers additional protections and most of the recent attacks haven't affected pnpm users.

## Release builds do not use dependency caches

Our CI for tests and on-demand builds runs entirely on GitHub, and we use caches because they save a lot of time. We make several PRs per day and waiting for every dependency to be downloaded from scratch on every branch would be painful.

Releases are different. They happen less often and they are what the user will receive. For releases we prefer the slower and cleaner path.

For release workflows we disable package manager caches:

```yaml
# rotki_release.yaml
enable-cache: false          # uv
package-manager-cache: false # pnpm
```

> The reason is cache poisoning. If a malicious dependency, binary, or intermediate artifact lands in a cache, the release job could reuse it without downloading the expected dependency again. That kind of failure is hard to notice because everything can still look reproducible from the outside.

Reviewer: ```suggestion
The reason is cache poisoning. If a malicious dependency, binary, or intermediate artifact lands in a cache, the release job could reuse it without ever pulling the legitimate version. That kind of failure is hard to notice because everything can still look reproducible from the outside.
```

## Build provenance attestations

Users should be able to ask: "Was this binary built by rotki, from the rotki source code, in the expected CI environment?"

To help answer that, we generate build provenance attestations for release artifacts using `actions/attest-build-provenance`. This attaches verifiable metadata to the artifact.

The attestation links the binary to the workflow that produced it and to the source revision used for the build. This is useful because it narrows the trust problem. Instead of only trusting a file that appeared on a release page, users and downstream distributors can verify where it came from.

This matters especially for desktop applications like rotki since people download a binary and run it locally. If an attacker manages to replace an artifact after the build, provenance checks give us and users another way to detect that the file is not the expected output of our release process.

More information is available [here](https://docs.rotki.com/requirement-and-installation/verify.html#verify-your-download)

> > Note: In the case of the Windows version we are forced to sign the binary locally with a yubikey and reupload it. The github attestation will fail due that.

Reviewer: ```suggestion
> Note: In the case of the Windows version we are forced to sign the binary locally with a yubikey and reupload it. The github attestation will fail because of that.
```

So we do not add dependencies casually. Before adopting a new package we look at things like:

- Who maintains it?
- Is the project active?
- Does it have a history of suspicious releases?
- How many transitive dependencies does it bring?
- Is the package doing something simple enough that we should implement it ourselves?
- Is vendoring a small piece of code safer than depending on the whole package?

When a dependency is accepted, it is always added with locked versions. We have to be honest about the fact that maintainers are themselves attack targets and that dependency graphs grow fast.

## Closing thoughts

Securing the supply chain for any software is hard and no single measure shields you. It takes several layered steps and even then nothing is perfect. The practices above are some of the most important ones we take to ensure that the software you receive from rotki is exactly what you expect from us.

> None of this means we are done. Security work is never finished. But we feel this is the right direction: less implicit trust, more reviewable state, and release artifacts that are easier to reason about.

Reviewer: ```suggestion
None of this means we are done. Security work is never finished. But we think this is the right direction: less implicit trust, more reviewable state, and release artifacts that are easier to reason about.
```
